### DISCUSSION TITLE
LeRes model for depth map

### DISCUSSION THREAD

**Discussion by bugmelone** [thygate/stable-diffusion-webui-depthmap-script#34] [Ideas]
Just wonder if it possibly to make?
Found this repo today:
https://github.com/compphoto/BoostingMonocularDepth
and i was surprised about level of details in output maps
![204601975-9b8f4b81-0f21-4dce-aac7-f01ff5a4c2dc](https://user-images.githubusercontent.com/114267100/204668005-2a2b0b58-1bb4-4259-acdb-ee967d90f9c3.jpg)
![204603285-41213326-7c88-4017-9831-4d29e29c8e7d](https://user-images.githubusercontent.com/114267100/204668019-7b2bb990-062a-4381-bf81-8e30446992c2.png)
![204603291-88ad3129-cddf-4068-811a-22e3452eb9fe](https://user-images.githubusercontent.com/114267100/204668023-8574d627-1e0a-4cd7-a697-f809ebba34af.png)

**Comment by bugmelone** [comment]
also you can try it yourself in their Colab
https://colab.research.google.com/github/compphoto/BoostingMonocularDepth/blob/main/Boostmonoculardepth.ipynb

**Comment by thygate** [comment]
Yes, I'm aware of the paper, I quote their observations in the readme. Your results are pretty amazing, I was planning to test their method in the extension, but did not notice LeRes was so much more detailed.

As a script this would have made installing too cumbersome, as an extension it is now possible to easy install more dependencies.

I will look into it.

**Comment by bugmelone** [comment]
Some conversation about LeRes model implementation in to the depthmap2mask extension. It might be interesting:
https://github.com/Extraltodeus/depthmap2mask/issues/11#issuecomment-1334722770

**Comment by thygate** [comment]
I'm working on it, the imports are a problem. I have the res101 net working now, just need to do some testing before i push it to the repo..

edit: it's live.

**Reply by bugmelone** [reply]
This is wonderful news! A big thank you for that research and for pushing your extension to the next level. I'm really glad of you and your work!

**Reply by AugmentedRealityCat** [reply]
I also want to personally thank you for making this happen.

**Reply by thygate** [reply]
You're very welcome.

**Comment by thygate** [comment]
You're very welcome, this is not yet multi-resolution merging as implemented by BoostingMonocularDepth, I will look into that next.

The intention/scope of this script was always to get the best possible depthmap and 3D related formats. This includes multi-resolution merging by BoostingMonocularDepth. I initially wasn't planning on including LeReS until I saw the results above.

Mesh and point-cloud generation are also future goals, including vt-vl-lab/3d-photo-inpainting. I could make those separate tabs together with BoostYourOwnDepth .. i'm not sure yet.

There's also a great fork by @graemeniedermayer [stable-diffusion-webui-normalmap-script](https://github.com/graemeniedermayer/stable-diffusion-webui-normalmap-script) focusing on normal maps with some very nice results already.

**Reply by AugmentedRealityCat** [reply]
Thanks a lot for pointing out that normal map script - I had completely missed it and I was in fact looking into exactly that kind of thing just yesterday to produce some low-relief stone carvings.

**Reply by graemeniedermayer** [reply]
Something super cool I recently found with lighting is you can use img2img to rebake in the new lighting from normal maps (this might work for depth maps too) and it seems to remove a good portion of the lighting artefacts. 

I don't mean to be self promo-y but here's my [quick explainer](https://www.youtube.com/watch?v=mrKbil_T36o ). I also did another mini-project using the projected depth maps in mobile augment reality and they look really cool!

Thank you thygate, this has opened up a lot of possibilities!

**Reply by bugmelone** [reply]
> Something super cool I recently found with lighting is you can use img2img to rebake in the new lighting from normal maps (this might work for depth maps too) and it seems to remove a good portion of the lighting artefacts.
> 
> I don't mean to be self promo-y but here's my [quick explainer](https://www.youtube.com/watch?v=mrKbil_T36o). I also did another mini-project using the projected depth maps in mobile augment reality and they look really cool!
> 
> Thank you thygate, this has opened up a lot of possibilities!

About your explanation, Wow, I didn't realize how deep we can get it. Thanks for inspire!

**Comment by thygate** [comment]
In my testing so far, res101 seems to generate more detail compared to dpt_large, but sometimes gets very wrong depth estimation for some objects, and doesn't always seem to recognize windows .. 

It still suffers from the same issue as the midas models, there is a trade-off between structural consistency and high-frequency details with respect to net size (see [observations](https://github.com/compphoto/BoostingMonocularDepth#observations)).  Multi-resolution merging is the way forward.

What are your experiences and thoughts, is LeReS really better overall ?

**Reply by graemeniedermayer** [reply]
That actually sounds excellent for normal maps!

**Reply by AugmentedRealityCat** [reply]
I am not satisfied with any of them. 

Yesterday I had to find a way to transform a 2d digital painting of a tree into a 3d model, and basically I used all the different algorithms I had access to to generate depthmaps - so I had 5 of them - and then I manually massaged them in photoshop to get what I wanted out of them before using them as source for displacement maps. What also helped a lot was that the digital painting I was working from was split in layers so I could work on each element separately and then reassemble them in 3d space. 

But it is my understanding as a user (so I might be completely wrong as I have no grasp on the programming side) that indeed multi-resolution merging is the way forward. 

Is LeRes better overall ? It is clearly more detailed, so if you are more into extracting relief information from a surface, like the bark from a tree, it is clearly better than anything else.

For some reason, the one that seems to give me the most "realistic" results is the DPT-Hybrid algorithm. I wish I could tell why, and why it doesn't at all on some images.

It reminds me a bit of the upscalers: some are better than others, but you'd be hard pressed to find the "absolute best" that would be ideal in all situations. And just like with upscalers, it might be best to use them in combination.

What about that [Structure-Guided-Ranking-Loss](https://github.com/KexianHust/Structure-Guided-Ranking-Loss) algo ? Is that being used anywhere ? Is it related to https://github.com/aim-uofa/AdelaiDepth that you implemented recently ?

One more question while I'm at it: is there any special advantage in the use of heatmaps instead of deptmaps ? Do they allow to encode more than 8 bit of depth information somehow ? I feel like this could be the case by not have all the RGB channels being always equal to each other, but I'm just guessing. Never used them before, except when I had to work with actual thermal cam footage, and even then we would set it to output greyscale instead.

**Reply by AugmentedRealityCat** [reply]
One more thing: [donlinglok](https://github.com/donlinglok) pointed out something that might be useful for merging LeRes and Midas:

https://github.com/vt-vl-lab/3d-photo-inpainting/issues/136

You might want to have a look at that.

**Comment by thygate** [comment]
>What about that [Structure-Guided-Ranking-Loss](https://github.com/KexianHust/Structure-Guided-Ranking-Loss) algo ?

The people at [BoostingMonocularDepth](https://github.com/compphoto/BoostingMonocularDepth) seem to be least excited about SGR, and it might be a pain to implement in the script, I have no intentions of adding it at the moment.

>One more question while I'm at it: is there any special advantage in the use of heatmaps instead of deptmaps ?

To make the depthmap easier to analyze for human eyes, Show HeatMap shows an extra image in the WebUI that has a color gradient applied. It is not saved.
It's easier for humans to detect tiny differences in a color gradient than it is for detecting brightness differences.

edit: messed up reply again.
Also keep in mind that these are all still just depth estimates, if some result has more detail, but isn't correct, it's not useful. It needs to be accurate too.

**Comment by bugmelone** [comment]
I tested v0.2.3 and cannot reproduce my previous results. I made img2img with denoising strength = 0 to remove the influence of the generative model, plus the DepthMap v0.2.3 script with res101 as the model. I adjusted the image size to match the original. Perhaps I used incorrect settings in this test, but I'm hoping someone can correct me or run the same test with their settings. 
![00042-2430792169-_depth](https://user-images.githubusercontent.com/114267100/205468393-d3e61699-01c0-42a8-ad2a-4669d2f77e5a.png)

**Reply by thygate** [reply]
Did you generate the previous results with BoostingMonocularDepth ? 
> this is not yet multi-resolution merging as implemented by BoostingMonocularDepth,

Leres is one of the models, next to midas and sgr that are used by BoostingMonocularDepth to generate multiple depthmaps that are then merged by another model.

**Reply by bugmelone** [reply]
Yes, with their notebook. Thanks for correcting me, now i was better understanding what do you mean about multi-resolution merging. I didn't catch that so close.

**Reply by thygate** [reply]
It's the next thing I will be looking at ..

**Reply by bugmelone** [reply]
Did you mean that? They call it MergeNet
https://github.com/compphoto/BoostYourOwnDepth

**Reply by thygate** [reply]
That's the stand-alone implementation of their merging operator, it allows for you to use only 2 depthmaps of different resolution that you provide and merge those. I might add support for this later in a separate tab or something, i'm not sure. I'm counting on the algorithm explained in BoostingMonocularDepth to select the regions and net sizes to generate the depthmaps at different resolutions to feed to the merging network to be the best possible output.

**Reply by bugmelone** [reply]
I hope this algorithm does not take too much of your time. Thanks for the conversation, and a big thank you for your time!

**Reply by thygate** [reply]
You're welcome. I'm taking a slight detour looking at the 3D reconstruction code in the LeReS repo. It can output a PLY point-cloud, and the results [here](https://github.com/aim-uofa/AdelaiDepth/tree/main/LeReS) look pretty good.

**Comment by thygate** [comment]
![tmpcopvxbxh](https://user-images.githubusercontent.com/96586282/205477901-175677d6-d631-4f89-8856-af32fedf4bfc.png)

I have the multi-resolution merging as implemented by BoostingMonocularDepth working, above was generated in the WebUI using leres and the image from bugmelone's start post.

Still have some things to clean up and test, but it's coming soon..

**Reply by thygate** [reply]
![tmpw93482bm](https://user-images.githubusercontent.com/96586282/205485551-138429f9-c082-4922-bc39-41dee3490f25.png)

**Reply by thygate** [reply]
it's live.

**Comment by thygate** [comment]
Multi-resolution merging is now implemented and significantly improves results, and res101 can now also compute on CPU in the new version.

Let's see some results =)

**Reply by graemeniedermayer** [reply]
This is insane! It's works super well on everything I've thrown at it so far! 

new
![00054-1892394486-_depth](https://user-images.githubusercontent.com/38518120/205537204-5d5fde50-23e9-4372-8c27-67ad674764a2.png)
old
![00078-3697106439-castle_depth](https://user-images.githubusercontent.com/38518120/205537254-aae104e3-e892-4769-bfab-2f2f84026cbf.png)

even the much more ambiguous scenes it does a great job at

new
![00052-4086006661-_depth](https://user-images.githubusercontent.com/38518120/205536490-90b1a5bb-8de0-4ae6-8335-26d90362ae6d.png)
old
![00080-1393410692-castle_depth](https://user-images.githubusercontent.com/38518120/205536518-4cc973b3-5e8c-4423-aba8-2576ec4b5fa3.png)
original
![00079-1393410692-castle](https://user-images.githubusercontent.com/38518120/205536589-df92ff44-0aea-4f96-af78-2cbcd1885171.png)

**Comment by thygate** [comment]
How about a 3D stereo/anaglyph image .. =)

Your picture looks pretty amazing, check here :  https://github.com/thygate/stable-diffusion-webui-depthmap-script/discussions/45

![anaglyph](https://user-images.githubusercontent.com/96586282/207832051-96044966-b809-4827-8298-81b53be348bd.jpg)

**Reply by bugmelone** [reply]
Wow, this looks really cool, it's time to watch it on the big screen! Thanks a lot!!!
By the way, I figured out the "boost rmax setting" yesterday. Raising the setting from 1600 to 3000, I was really impressed with the level of detail, it turned out even cooler than in my first post here. I really appreciate the ability to change this setting! 
Also, thanks for the separate tab, very useful!

**Reply by thygate** [reply]
ah yes, i forgot adding an explanation to the readme for the setting. The default value is 3000 indeed, it will determine the max "wholeimage size" which sets the sizes for the intermediate depth maps it generates. I set it to 1600 default so people with little VRAM can run it too.
2400 seems te be max for boost with res101 on 8GB vram .. dpt_large might not be even to do that..

And you're very welcome, the separate tab was the most requested feature for simpler processing of existing images.
